### PR TITLE
feat: add support for custom SEO per page

### LINKED ISSUE #814: Feature: Microformat SEO data per page
### Type of feature

🍕 Feature

### Current behavior

We are not making User profiles and Insight Pages friendly with SEO. 

related, https://github.com/open-sauced/insights/issues/687

### Suggested solution

Set up search engine enrichment data. Not sure what the solution is in next.js, but would love thoughts. 

Here is an article for Next.js - https://nextjs.org/learn/seo/introduction-to-seo

Context from Google - https://developers.google.com/search/docs/appearance/structured-data/search-gallery

### Additional context

cc @0-vortex 

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Contributing Docs

- [ ] I agree to follow this project's Contribution Docs

### PR DESCRIPTION
**Pull request by Deadreyo**
## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Description

- Implements a new system to ease adding SEO to pages, and provides default SEO data if not specified.
- Fills in SEO data for the pages.

Technical implementation details present in a comment below.

## Related Tickets & Documents

part of #814 
### User pages:
image

## Added tests?

- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [ ] 🙅 no documentation needed

### CODE REVIEW

File: pages/user/[username].tsx [hunk 1]
```diff
@@ -26,23 +27,41 @@ const Contributor = (): JSX.Element => {
...
-        openPrs={contributor[0]?.recent_opened_prs}
-        recentContributionCount={contributor[0]?.recent_contribution_count}
-        prReviews={contributor[0]?.recent_pr_reviews}
-        prVelocity={contributor[0]?.recent_pr_velocity}
-      />
-    </div>
+    <>
+      <Head>
+        <title key="title">{contributorLogin} | Open Source Contributors</title>
+        <meta name="description" content={`Profile Bio: ${user?.bio || "I am an open source developer with a passion for music and video games. I strive to improve the open source community and am always looking for new ways to contribute."}`} key="description" />
```
Comment (bdougie) [inline pages/user/[username].tsx:33]:
```suggestion
        <meta name="description" content={`${user?.bio || "I am an open source developer with a passion for music and video games. I strive to improve the open source community and am always looking for new ways to contribute."}`} key="description" />
```

I'd remove "Profile Bio:"

File: pages/user/[username].tsx [hunk 2]
```diff
@@ -26,23 +27,41 @@ const Contributor = (): JSX.Element => {
...
+        <meta name="keywords" content="open source, github, contributor, profile, portfolio, developer, software engineer, software developer, software engineer portfolio, software developer portfolio, open source portfolio, open source developer, developer portfolio" key="keywords" />
+
+        <meta property="og:description" content={`Profile Bio: ${user?.bio || "I am an open source developer with a passion for music and video games. I strive to improve the open source community and am always looking for new ways to contribute."}`} key="og:description" />
```
Comment (bdougie) [inline pages/user/[username].tsx:36]:
```suggestion
        <meta property="og:description" content={`${user?.bio || "I am an open source developer with a passion for music and video games. I strive to improve the open source community and am always looking for new ways to contribute."}`} key="og:description" />
```

File: pages/user/[username].tsx [hunk 3]
```diff
@@ -26,23 +27,41 @@ const Contributor = (): JSX.Element => {
...
-        prTotal={contributor[0]?.recent_pr_total}
-        openPrs={contributor[0]?.recent_opened_prs}
-        recentContributionCount={contributor[0]?.recent_contribution_count}
-        prReviews={contributor[0]?.recent_pr_reviews}
-        prVelocity={contributor[0]?.recent_pr_velocity}
-      />
-    </div>
+    <>
+      <Head>
+        <title key="title">{contributorLogin} | Open Source Contributors</title>
```
Comment (bdougie) [inline pages/user/[username].tsx:32]:
```suggestion
        <title key="title">{contributorLogin} | OpenSauced</title>
```

File: pages/user/[username].tsx [hunk 4]
```diff
@@ -26,23 +27,41 @@ const Contributor = (): JSX.Element => {
...
+        <meta name="description" content={`Profile Bio: ${user?.bio || "I am an open source developer with a passion for music and video games. I strive to improve the open source community and am always looking for new ways to contribute."}`} key="description" />
+        <meta name="keywords" content="open source, github, contributor, profile, portfolio, developer, software engineer, software developer, software engineer portfolio, software developer portfolio, open source portfolio, open source developer, developer portfolio" key="keywords" />
+
+        <meta property="og:description" content={`Profile Bio: ${user?.bio || "I am an open source developer with a passion for music and video games. I strive to improve the open source community and am always looking for new ways to contribute."}`} key="og:description" />
+        <meta property="og:image" content={profile.githubAvatar} key="og:image" />
+        <meta property="og:site_name" content="OpenSauced Insights" key="og:site_name" />
+        <meta property="og:type" content="website" key="og:type" />
+
+        <meta name="twitter:title" content={`${contributorLogin} | Open Source Contributors`} key="twitter:title" />
```
Comment (bdougie) [inline pages/user/[username].tsx:41]:
```suggestion
        <meta name="twitter:title" content={`${contributorLogin} | OpenSauced`} key="twitter:title" />
```

**Comment by OgDev-01** [comment]
IMHO, Using an SEO wrapper should be best, to avoid too many duplicates here and there 💯, I have a rough example here https://github.com/OgDev-01/ecommerce-frontend/blob/main/src/components/organisms/Layouts/Layout.jsx 🍕

**Comment by bdougie** [comment]
> IMHO, Using an SEO wrapper should be best, to avoid too many duplicates here and there 💯, I have a rough example here https://github.com/OgDev-01/ecommerce-frontend/blob/main/src/components/organisms/Layouts/Layout.jsx 🍕

I can't speak to the technical implementation, but can say this works for me as a first pass. SEO is something that we will continue to work on, but we are starting at 0 and the sooner we can figure this out the better.

**Comment by Deadreyo** [comment]
Explanation of the new SEO implementation:
- Similar to PageLayout, an "SEO" object can be assigned to the component that dictates the information to be put in the metadata.
- If no SEO object is passed, default data will be used.
Example:
image

- Dynamic SEO changes is also supported by a method "updateSEO" that is assigned to the rendered page component in `_app.tsx`.
- This method can be used to change the metadata while inside the component.
Example:
image
- This would of course override the static SEO object if present.

Feedback appreciated!

**Comment by brandonroberts** [comment]
@Deadreyo I think its enough to move forward with. We can revisit if/when we move to the `app` directory. in Next 13. If you'll clean up the build failures, we can get it merged into beta

**Review by bdougie — APPROVED** [review]

**Review by brandonroberts — APPROVED** [review]

**Pull request merged**